yi-core/src/Yi/CompletionTree.hs

```diff
@@ -1,59 +1,158 @@
--- | This is a little helper for completion interfaces.
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
 
-module Yi.CompletionTree (
-        CompletionTree(CT),
-        stepTree, obvious, mergeTrees, listToTree, complete)
-   where
+{- |
+Module      :  Yi.CompletionTree
+License     :  GPL-2
+Maintainer  :  yi-devel@googlegroups.com
+Stability   :  experimental
+Portability :  portable
 
-import Control.Arrow (first)
-import Data.List     (find, groupBy, intercalate, sortBy)
+Little helper for completion interfaces.
 
--- inside a completion tree, the a's must be unique on each level
+Intended to be imported qualified:
 
-data CompletionTree a = CT [(a,CompletionTree a)]
+>import qualified Yi.CompletionTree as CT
+-}
+module Yi.CompletionTree (
+  -- * CompletionTree type
+  CompletionTree (),
+  -- * Lists
+  fromList, toList,
+  -- * Modification
+  complete, update, updateOne,
+  -- * Debugging
+  pretty
+  ) where
 
-instance (Show a) => Show (CompletionTree a) where
-  show = show'
+import           Control.Arrow   (first)
+import           Data.Function   (on)
+import           Data.List       (partition, maximumBy, intercalate)
+import qualified Data.Map.Strict as M
+import           Data.Map.Strict (Map)
+import           Data.Maybe      (isJust, fromJust, listToMaybe, catMaybes)
+import qualified Data.ListLike   as LL
+import           Data.ListLike   (ListLike)
 
-show' :: (Show a) => CompletionTree a -> String
-show' (CT []) = []
-show' (CT [(a,st)]) = shows a $ show' st
-show' (CT trees) = "[" ++ intercalate "|" (map (\(x,y)->shows x $ show' y) trees) ++ "]"
+-- | A CompletionTree is a map of partial completions.
+newtype CompletionTree a = CompletionTree (Map a (CompletionTree a)) deriving (Monoid)
 
-compareBy :: (Ord b) => (a->b)->a->a->Ordering
-compareBy f a b = compare (f a) (f b)
+instance (Show a, ListLike a i) => Show (CompletionTree a) where
+  show ct = "fromList " ++ show (toList ct)
 
-listToTree :: [a] -> CompletionTree a
-listToTree = foldr (\a b->CT [(a,b)]) (CT [])
+stripPrefix :: (Eq item, ListLike full item) => full -> full -> Maybe full
```

Comment:
You've made this function pretty efficient for a list. But we don't care about performance on lists as much as text, and now its slow on Text. I'd recommend using `if a \`isPrefixOf\` b`